CI/COMPAT: read_stata failing in numpy dev pipeline

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Problem description

Looks like a change in numpy broke our `read_stata` code, so now we have a bunch of `read_stata` test errors popping up during CI in the numpy_dev pipeline. Maybe this [Numpy PR](https://github.com/numpy/numpy/pull/16943) somehow broke things? I don't think any other PR merged during the last 24 hours could have possibly done so, although I might be missing something.

Error example:
```
pandas/tests/io/test_stata.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/stata.py:1928: in read_stata
    data = reader.read()
pandas/io/stata.py:1646: in read
    cols_ = np.where(self.dtyplist)[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ([<class 'numpy.int8'>, <class 'numpy.int16'>, <class 'numpy.int32'>, <class 'numpy.float32'>, <class 'numpy.float64'>, <class 'numpy.float32'>, ...],)
kwargs = {}
relevant_args = ([<class 'numpy.int8'>, <class 'numpy.int16'>, <class 'numpy.int32'>, <class 'numpy.float32'>, <class 'numpy.float64'>, <class 'numpy.float32'>, ...], None, None)

>   ???
E   ValueError: invalid __array_struct__
```

--- a/pandas/io/stata.py
+++ b/pandas/io/stata.py
@@ -1643,8 +1643,7 @@ the string values returned are correct."""
 
         data = self._insert_strls(data)
 
-        cols_ = np.where(self.dtyplist)[0]
-
+        cols_ = np.where([dtyp is not None for dtyp in self.dtyplist])[0]
         # Convert columns (if needed) to match input type
         ix = data.index
         requires_type_conversion = False

